### PR TITLE
system-variables: add fine_grained_shuffle

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)

Add two system vars description.
### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [ ] v6.2 (TiDB 6.2 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: system-variables.md [hunk 1]
```diff
@@ -2051,3 +2051,22 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+### `tiflash_fine_grained_shuffle_stream_count` <span class="version-mark">从 v6.2.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`-1`
+- 范围：`[-1, 1024]`
+- 当 WindowFunction 下推到 TiFlash 执行时，可以通过该变量控制 WindowFunction 执行的并行度。不同取值含义：
+
+    * 0: 表示不使用细粒度 shuffle 功能，下推到 TiFlash 的 WindowFunction 以单线程方式执行
+    * 大于 0：表示使用细粒度 shuffle 功能，下推到 TiFlash 的 WindowFunction 会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
+    * -1: 表示使用细粒度 shuffle 功能。如果 `tidb_max_tiflash_threads` 有效（大于 0），则 `tiflash_fine_grained_shuffle_stream_count` 会自动取值为 `tidb_max_tiflash_threads` ，否则为默认值 8 。最终在 TiFlash 上 WindowFunction 的实际并发度为：min(`tiflash_fine_grained_shuffle_stream_count`,     TiFlash 节点物理线程数)
```
Comment (hfxsd) [inline system-variables.md:2064]:
```suggestion
    * -1: 表示使用细粒度 shuffle 功能。如果 `tidb_max_tiflash_threads` 有效（大于 0），则 `tiflash_fine_grained_shuffle_stream_count` 会自动取值为 `tidb_max_tiflash_threads` ，否则为默认值 8 。最终在 TiFlash 上 WindowFunction 的实际并发度为：min(`tiflash_fine_grained_shuffle_stream_count`，TiFlash 节点物理线程数)
```
```suggestion
    * -1: 表示使用细粒度 shuffle 功能。如果 `tidb_max_tiflash_threads` 有效（大于 0），则 `tiflash_fine_grained_shuffle_stream_count` 会自动取值为 `tidb_max_tiflash_threads` ，否则为默认值 8 。最终在 TiFlash 上 WindowFunction 的实际并发度为：min(`tiflash_fine_grained_shuffle_stream_count`,     TiFlash 节点物理线程数)
```

File: system-variables.md [hunk 2]
```diff
@@ -2051,3 +2051,22 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+- 建议：WindowFunction 的性能应该随着该值的增加线性提升。但是如果设置过大，超过实际的物理线程数，反而会导致性能下降。
+
+### `tiflash_fine_grained_shuffle_batch_size` <span class="version-mark">从 v6.2.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`8192`
+- 细粒度 shuffle 功能开启时，该变量控制发送端发送数据的攒批大小，即发送端累计行数超过该值就会进行一次数据发送。
+- 对性能影响：该值设置过小，会导致数据发送过于频繁，降低向量化效果。如果设置过大会导致发送频率过低，接收端大部分时间在等待数据，影响性能。
```
Comment (hfxsd) [inline system-variables.md:2072]:
```suggestion
- 对性能影响：请根据业务情况设置合理的大小，设置不当会影响性能。如果该值设置过小，会导致数据发送过于频繁，降低向量化效果。如果设置过大，会导致发送频率过低，接收端大部分时间在等待数据，影响性能。
```

File: system-variables.md [hunk 3]
```diff
@@ -2051,3 +2051,22 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`-1`
+- 范围：`[-1, 1024]`
+- 当 WindowFunction 下推到 TiFlash 执行时，可以通过该变量控制 WindowFunction 执行的并行度。不同取值含义：
+
+    * 0: 表示不使用细粒度 shuffle 功能，下推到 TiFlash 的 WindowFunction 以单线程方式执行
+    * 大于 0：表示使用细粒度 shuffle 功能，下推到 TiFlash 的 WindowFunction 会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
+    * -1: 表示使用细粒度 shuffle 功能。如果 `tidb_max_tiflash_threads` 有效（大于 0），则 `tiflash_fine_grained_shuffle_stream_count` 会自动取值为 `tidb_max_tiflash_threads` ，否则为默认值 8 。最终在 TiFlash 上 WindowFunction 的实际并发度为：min(`tiflash_fine_grained_shuffle_stream_count`，TiFlash 节点物理线程数)
+- 建议：WindowFunction 的性能应该随着该值的增加线性提升。但是如果设置过大，超过实际的物理线程数，反而会导致性能下降。
```
Comment (hfxsd) [inline system-variables.md:2065]:
```suggestion
- 建议：WindowFunction 的性能应该随着该值的增加线性提升。但是如果设置的值超过实际的物理线程数，反而会导致性能下降。
```

File: system-variables.md [hunk 4]
```diff
@@ -2051,3 +2051,23 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+### `tiflash_fine_grained_shuffle_stream_count` <span class="version-mark">从 v6.2.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`-1`
+- 范围：`[-1, 1024]`
+- 当 WindowFunction 下推到 TiFlash 执行时，可以通过该变量控制 WindowFunction 执行的并行度。不同取值含义：
+
+    * 0: 表示不使用细粒度 shuffle 功能，下推到 TiFlash 的 WindowFunction 以单线程方式执行
+    * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的 WindowFunction 会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
+    * -1: 表示使用细粒度 shuffle 功能。如果 `tidb_max_tiflash_threads` 有效（大于 0），则 `tiflash_fine_grained_shuffle_stream_count` 会自动取值为 `tidb_max_tiflash_threads` ，否则为默认值 8 。最终在 TiFlash 上 WindowFunction 的实际并发度为：min(`tiflash_fine_grained_shuffle_stream_count`，TiFlash 节点物理线程数)
```
Comment (hfxsd) [inline system-variables.md:2064]:
```suggestion
    * 0: 表示不使用细粒度 shuffle 功能，下推到 TiFlash 的 WindowFunction 以单线程方式执行。
    * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的 WindowFunction 会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)。
    * -1: 表示使用细粒度 shuffle 功能。如果 `tidb_max_tiflash_threads` 有效（大于 0），则 `tiflash_fine_grained_shuffle_stream_count` 会自动取值为 `tidb_max_tiflash_threads` ，否则为默认值 8 。最终在 TiFlash 上 WindowFunction 的实际并发度为：min(`tiflash_fine_grained_shuffle_stream_count`，TiFlash 节点物理线程数)。
```

File: system-variables.md [hunk 5]
```diff
@@ -2051,3 +2051,23 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
 - 是否持久化到集群：是
 - 默认值：`ON`
 - 这个变量用于控制计算窗口函数时是否采用高精度模式。
+
+### `tiflash_fine_grained_shuffle_stream_count` <span class="version-mark">从 v6.2.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`-1`
+- 范围：`[-1, 1024]`
+- 当 WindowFunction 下推到 TiFlash 执行时，可以通过该变量控制 WindowFunction 执行的并行度。不同取值含义：
```
Comment (ran-huang) [inline system-variables.md:2060]: WindowFunction 可以直接写窗口函数吗

Comment (ran-huang) [reply system-variables.md:2060]: 建议把变量按字母顺序排列

File: system-variables.md [hunk 6]
```diff
@@ -2051,3 +2051,23 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+
+    * 0: 表示不使用细粒度 shuffle 功能，下推到 TiFlash 的 WindowFunction 以单线程方式执行。
+    * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的 WindowFunction 会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)。
+    * -1: 表示使用细粒度 shuffle 功能。如果 `tidb_max_tiflash_threads` 有效（大于 0），则 `tiflash_fine_grained_shuffle_stream_count` 会自动取值为 `tidb_max_tiflash_threads` ，否则为默认值 8 。最终在 TiFlash 上 WindowFunction 的实际并发度为：min(`tiflash_fine_grained_shuffle_stream_count`，TiFlash 节点物理线程数)。
```
Comment (ran-huang) [inline system-variables.md:2064]: `tidb_max_tiflash_threads` 的链接可以附上

Comment (hfxsd) [inline system-variables.md:2064]:
```suggestion
    * -1: 表示使用细粒度 shuffle 功能。如果 [`tidb_max_tiflash_threads`](/system-variables.md#tidb_max_tiflash_threads-从-v610-版本开始引入) 有效（大于 0），则 `tiflash_fine_grained_shuffle_stream_count` 会自动取值为 `tidb_max_tiflash_threads` ，否则为默认值 8 。最终在 TiFlash 上 WindowFunction 的实际并发度为：min(`tiflash_fine_grained_shuffle_stream_count`，TiFlash 节点物理线程数)。
```

File: system-variables.md (hunk 5 above)
Comment (hfxsd) [reply system-variables.md:2060]: @guo-shaoge 麻烦看下？

Comment (guo-shaoge) [reply system-variables.md]: done

File: system-variables.md (hunk 6 above)
Comment (guo-shaoge) [reply system-variables.md]: done

**Review by ran-huang — APPROVED** [review]

**Review by hfxsd — APPROVED** [review]
LGTM

File: system-variables.md [hunk 7]
```diff
@@ -1980,6 +1980,26 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`-1`
+- 范围：`[-1, 1024]`
+- 当窗口函数下推到 TiFlash 执行时，可以通过该变量控制窗口函数执行的并行度。不同取值含义：
+
+    * 0: 表示不使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数以单线程方式执行
+    * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
+    * -1: 表示使用细粒度 shuffle 功能。如果 [`tidb_max_tiflash_threads`](/system-variables.md#tidb_max_tiflash_threads-从-v610-版本开始引入) 有效（大于 0），则 `tiflash_fine_grained_shuffle_stream_count` 会自动取值为 [`tidb_max_tiflash_threads`](/system-variables.md#tidb_max_tiflash_threads-从-v610-版本开始引入) ，否则为默认值 8 。最终在 TiFlash 上窗口函数的实际并发度为：min(`tiflash_fine_grained_shuffle_stream_count`，TiFlash 节点物理线程数)
+- 建议窗口函数的性能应该随着该值的增加线性提升。但是如果设置的值超过实际的物理线程数，反而会导致性能下降。
```
Comment (guo-shaoge) [inline system-variables.md:2001]:
```suggestion
- 理论上窗口函数的性能会随着该值的增加线性提升。但是如果设置的值超过实际的物理线程数，反而会导致性能下降。
```

File: system-variables.md [hunk 8]
```diff
@@ -1980,6 +1980,26 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
 - 默认值 `-1` 表示使用 `tidb_executor_concurrency` 的值。
 
+### `tiflash_fine_grained_shuffle_batch_size` <span class="version-mark">从 v6.2.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`8192`
+- 范围：`[1, 18446744073709551616]`
+- 细粒度 shuffle 功能开启时，该变量控制发送端发送数据的攒批大小，即发送端累计行数超过该值就会进行一次数据发送。
+- 对性能影响：请根据业务情况设置合理的大小，设置不当会影响性能。如果该值设置过小，会导致数据发送过于频繁，降低向量化效果。如果设置过大，会导致发送频率过低，接收端大部分时间在等待数据，影响性能。
+
```
Comment (moyun) [inline system-variables.md:1990]: 这段话稍微修改一下吧，这么写别人比较难理解到底应该怎么设置，那个取值域那么宽，都是有意义的吗？假如取极端值会有什么具体的意义？
建议这么描述改变这个值的效果： 缩小此数值，可以导致`正反两方面的什么样的影响`；此值增大，可导致`正反两方面的什么样的影响`。建议用户`如何调整`此数值（给一个怎么样观测效果调整数值方法）。

File: system-variables.md [hunk 9]
```diff
@@ -1980,6 +1980,26 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+### `tiflash_fine_grained_shuffle_stream_count` <span class="version-mark">从 v6.2.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`-1`
+- 范围：`[-1, 1024]`
+- 当窗口函数下推到 TiFlash 执行时，可以通过该变量控制窗口函数执行的并行度。不同取值含义：
+
+    * 0: 表示不使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数以单线程方式执行
+    * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
+    * -1: 表示使用细粒度 shuffle 功能。如果 [`tidb_max_tiflash_threads`](/system-variables.md#tidb_max_tiflash_threads-从-v610-版本开始引入) 有效（大于 0），则 `tiflash_fine_grained_shuffle_stream_count` 会自动取值为 [`tidb_max_tiflash_threads`](/system-variables.md#tidb_max_tiflash_threads-从-v610-版本开始引入) ，否则为默认值 8 。最终在 TiFlash 上窗口函数的实际并发度为：min(`tiflash_fine_grained_shuffle_stream_count`，TiFlash 节点物理线程数)
```
Comment (moyun) [inline system-variables.md:2000]: 这几个数值代表的意义不太合乎常规习惯吧？ -1 == enable， 0 == disable ？ ... > 0 又代表 enable && parallelism, 这是不是混乱了点？ disable 和 enable 映射到取值范围上还不连续呀？

**Review by moyun — CHANGES_REQUESTED** [review]

File: system-variables.md (hunk 9 above)
Comment (guo-shaoge) [reply system-variables.md:2000]: working on it:https://github.com/pingcap/tidb/pull/36621

File: system-variables.md [hunk 10]
```diff
@@ -1980,6 +1980,26 @@ Query OK, 0 rows affected, 1 warning (0.00 sec)
...
+
+### `tiflash_fine_grained_shuffle_stream_count` <span class="version-mark">从 v6.2.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`0`
+- 范围：`[-1, 1024]`
+- 当窗口函数下推到 TiFlash 执行时，可以通过该变量控制窗口函数执行的并行度。不同取值含义：
+
+    * -1: 表示不使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数以单线程方式执行
+    * 大于 0: 表示使用细粒度 shuffle 功能，下推到 TiFlash 的窗口函数会以多线程方式执行，并发度为： min(`tiflash_fine_grained_shuffle_stream_count`, TiFlash 节点物理线程数)
```
Comment (moyun) [inline system-variables.md:1999]: 描述的顺序可否调整一下？

1. `-1`， 
2. `0`， 
3. `>0`

Comment (guo-shaoge) [reply system-variables.md]: done

**Comment by moyun** [comment]
LGTM

**Review by moyun — APPROVED** [review]

**Pull request merged**